common/src/chain/transaction/signature.rs

```diff
@@ -0,0 +1,512 @@
+                SigHashType::try_from(SigHashType::NONE | SigHashType::ANYONECANPAY).unwrap(),
+                vec![1, 2, 3, 4, 5, 6, 7, 8, 9],
+            )),
+        )
+        .unwrap();
+        assert_eq!(
+            verify_signature(&outpoint_dest, &tx, 0),
+            Err(TransactionSigError::InvalidSignatureEncoding)
+        );
+
+        // SigHashType SINGLE - must fail because there are no bytes in raw_signature
+        tx.update_witness(
+            0,
+            InputWitness::Standard(StandardInputSignature::new(
+                SigHashType::try_from(SigHashType::SINGLE).unwrap(),
+                vec![],
+            )),
+        )
+        .unwrap();
+        assert_eq!(
+            verify_signature(&outpoint_dest, &tx, 0),
+            Err(TransactionSigError::InvalidSignatureEncoding)
+        );
+
+        // SigHashType SINGLE - must fail because there are wrong bytes in raw_signature
+        tx.update_witness(
+            0,
+            InputWitness::Standard(StandardInputSignature::new(
+                SigHashType::try_from(SigHashType::SINGLE | SigHashType::ANYONECANPAY).unwrap(),
+                vec![1, 2, 3, 4, 5, 6, 7, 8, 9],
+            )),
+        )
+        .unwrap();
+        assert_eq!(
+            verify_signature(&outpoint_dest, &tx, 0),
+            Err(TransactionSigError::InvalidSignatureEncoding)
+        );
+    }
+
+    #[test]
+    #[allow(clippy::eq_op)]
+    fn check_invalid_input_index_for_verify_signature() {
+        const INVALID_INPUT_INDEX: usize = 1234567890;
+
+        let (private_key, public_key) = PrivateKey::new(KeyKind::RistrettoSchnorr);
+        let outpoint_dest = Destination::PublicKey(public_key);
+        let sighash_type = SigHashType::try_from(SigHashType::ALL).unwrap();
+        let mut tx = generate_unsign_tx(outpoint_dest.clone()).unwrap();
+        sign_tx(&mut tx, &private_key, sighash_type, outpoint_dest.clone());
+        // input index out of range
+        assert_eq!(
+            verify_signature(&outpoint_dest, &tx, INVALID_INPUT_INDEX),
+            Err(TransactionSigError::InvalidInputIndex(
+                INVALID_INPUT_INDEX,
+                1
+            ))
+        );
+    }
+
+    #[test]
+    fn sign_modify_then_verify() {
+        // Create and sign tx, and then modify and verify it.
+        let (private_key, public_key) = PrivateKey::new(KeyKind::RistrettoSchnorr);
+        let outpoint_dest = Destination::PublicKey(public_key);
+        let sighash_type = SigHashType::try_from(SigHashType::ALL).unwrap();
+        let mut original_tx = generate_unsign_tx(outpoint_dest.clone()).unwrap();
+        sign_tx(
+            &mut original_tx,
+            &private_key,
+            sighash_type,
+            outpoint_dest.clone(),
+        );
+        assert_eq!(verify_sign_tx(&original_tx, &outpoint_dest), Ok(()));
+
+        // Should failed due to changed flags
+        let mut tx_updater = TransactionUpdater::try_from(&original_tx).unwrap();
```

Comment:
I'm having trouble understanding the benefit of `TransactionUpdater`... if I understand correctly, it's just a fancy `clone()`... why not just `original_tx.clone()` to a mutable and then modify it?!

Reply:
So, let me try to explain. We can't mutate `Transaction` or `TransactionV1` due to they have private fields. We have two ways to solve it:
1. We can add functions that will mutate inner state (tx.set_flag \ tx.set_inputs \ etc)
2. We can add a type with public fields. We load inner data into this type, mutable it, and then create a new tx. For me, this way looks better.

Reply:
I see. Alright. Keep that for now. Just add a comment above `TransactionUpdater` saying "this is required because we can't access private fields of a the `Transaction` class". Also please call it `MutableTransaction`.